docs/src/explanations/Documentation-of-functions.md

```diff
@@ -0,0 +1,316 @@
+---
+
+layout: docs
+
+title:  "Documention of functions"
+
+section: "chisel3"
+
+---
+
+# Further Explanation for the := and  <> Operators
+
+Chisel contains two connection operators, `:=` and `<>`. This document serves for a deeper explanation of the differences of the two and when to use one or the other. The differences are demonstrated with experiments using Scastie examples + DecoupledIO examples.
+
+
+
+### Experiment Setup
+
+```scala mdoc
+// Imports used by the following examples
+import chisel3._
+import chisel3.util.DecoupledIO
+```
+
+The diagram for the experiment can be viewed **[here](https://docs.google.com/document/d/14C918Hdahk2xOGSJJBT-ZVqAx99_hg3JQIq-vaaifQU/edit?usp=sharing)**
+![Experiment Image](https://raw.githubusercontent.com/chipsalliance/chisel3/master/docs/src/images/chisel_01.png?sanitize=true)
+
+```scala mdoc:silent
+
+class Wrapper extends Module{
+  val io = IO(new Bundle {
+  val in = Flipped(DecoupledIO(UInt(8.W)))
+  val out = DecoupledIO(UInt(8.W))
+  })
+  val p = Module(new PipelineStage)
+  val c = Module(new PipelineStage) 
+  // connect Producer to IO
+  p.io.a <> io.in
+  // connect producer to consumer
+  c.io.a <> p.io.b
+  // connect consumer to IO
+  io.out <> c.io.b
+}
+class PipelineStage extends Module{
+  val io = IO(new Bundle{
+    val a = Flipped(DecoupledIO(UInt(8.W)))
+    val b = DecoupledIO(UInt(8.W))
+  })
+  io.b <> io.a
+}
+```
+Below we can see the resulting Verilog for this example:
+```scala modc
+ChiselStage.emitVerilog(new Wrapper)
+```
+## Concept 1: <> is Communicative
+
+
+
+This experiment is set up to test for the function of `<>` using the experiment above.
+
+Achieving this involves flipping the RHS and LHS of each arrow and seeing how '<>'  will react.
+( Scastie link for the experiment:https://scastie.scala-lang.org/Shorla/LVhlbkFQQnq7X3trHfgZZQ )
+
+
+
+
+```scala mdoc:silent:reset
+import chisel3._
+import chisel3.util.DecoupledIO
+
+class Wrapper extends Module{
+  val io = IO(new Bundle {
+  val in = Flipped(DecoupledIO(UInt(8.W)))
+  val out = DecoupledIO(UInt(8.W))
+  })
+  val p = Module(new PipelineStage)
+  val c = Module(new PipelineStage) 
+  // connect Producer to IO
+  io.in := p.io.a
+  // connect producer to consumer
+  p.io.b := c.io.a
+  // connect consumer to IO
+  c.io.b := io.out
+}
+
+class PipelineStage extends Module{
+  val io = IO(new Bundle{
+    val a = Flipped(DecoupledIO(UInt(8.W)))
+    val b = DecoupledIO(UInt(8.W))
+  })
+  io.b <> io.a
+}
+```
+Below we can see the resulting Verilog for this example:
+```scala modc
+ChiselStage.emitVerilog(new Wrapper)
+```
+### Conclusion: 
+The Verilog remained the same without incurring errors, showing that the `<>` operator is generally communicative.
+
+
+
+
+## Concept 2: := means assign ALL signals from the RHS to the LHS, regardless of their direction.
```

Comment:
This is not true. The RHS can have more `Record` fields than the LHS; in this case, only the RHS elements that can be mapped to LHS are assigned.

Reply:
@debs-sifive so not all signals From the RHS are assigned to the LHS, Does the experiment on concept 2 show that?

Reply:
No, since you're connecting two of the same kind of bundle. The RHS bundle can have more elements in it than the LHS.

Reply:
I think we need a different, simpler experiment that just shows this for single-directional bundle, where one side has an `a`, `b`, and `c` and the other only has `a`, `b`.... and show what happens when they are on either side of the `:=`

Reply:
Okay, but will be new experiment still be in this document or an entire new document will have to be created for that?

Reply:
> The RHS bundle can have more elements in it than the LHS.

@debs-sifive that is not what my experimenting shows in https://scastie.scala-lang.org/MOR2DI4KT9KxElvXBWYiWA ... the error is `chisel3.internal.ChiselException: Connection between left (FilterIO(IO in unelaborated Block)) and source (NotReallyAFilterIO(IO in unelaborated Block)) failed @.z: Left Record missing field (z).`

Are you sure about the statement? Or is what you are saying true only for `:=`, not `<>`?

Reply:
Confirmed that it gives similar error for `:=`

Reply:
> Okay, but will the new experiment still be in this document or an entirely new document will have to be created for that?

@Shorla  the simpler document is in the existing code about "Interfaces and Connections" that we discussed yesterday

Reply:
There is some confusion in this comment thread between this concept 2 and this concept 7. This is about ignoring direction, where concept 7 is about signals being present or not. Let's keep this example about the directionality aspect.

Reply:
Coercing inputs to inputs does not work: https://scastie.scala-lang.org/reYHtvBMS32xK4O6mj7xLg

What about outputs... does work: https://scastie.scala-lang.org/rSRU7as8SQC7MG8vevzliQ
Need to check th eresult of that and see what it generated...

Reply:
@mwachs5 my comment was just referring to the `:=` MonoConnect. I'm going off the documentation here: https://github.com/chipsalliance/chisel3/blob/849d4a0b7f6f7ea056c5280b9d319dadf5225022/core/src/main/scala/chisel3/internal/MonoConnect.scala#L22

This example demonstrates a compiling `:=` connection with more elements on the RHS than LHS: https://scastie.scala-lang.org/opTU5MznT5OmQhIIsYUylA

The `<>` BiConnect is expected to error if Bundle elements do not match up exactly, as documented here:
https://github.com/chipsalliance/chisel3/blob/849d4a0b7f6f7ea056c5280b9d319dadf5225022/core/src/main/scala/chisel3/internal/BiConnect.scala#L23

Reply:
Ok. I am realizing that a lot of confusion is because this section is called "Bulk Connect". but `:=` is a bulk connect just as much as `<>` is. I think we should rename this header from "Bulk Conections" to "Bidirectional Connections `<>`"